### PR TITLE
Migrate wasm-themis to TypeScript

### PR DESCRIPTION
**Pull request by maxammann**
So my PR for TS is finally here. This PR only changes syntax and should not change semantics. This is not possible in every case, but in all cases that matter.

The module works in node, older browsers and also works using ES6 and TypeScript. Here is an example for node:

```
mkdir test && cd test
npm init -y
npm install file://./wasm-themis-0.14.0.tgz
echo "const themis = require('wasm-themis'); \
themis.initialize().then(() => { \
    const cell = themis.SecureCellSeal.withPassphrase('pass'); \
    console.log(cell.encrypt(new Uint8Array([1]))); \
})" > example.js
node example.js
```

Example for web like a react app:

```ts
// @ts-ignore
import themisWasm from "wasm-themis/dist/libthemis.wasm";
import { initialize, SecureCellSeal } from "wasm-themis"; // webpack takes care of making the wasm file available

await initialize(themisWasm);
const cell = SecureCellSeal.withPassphrase(pw);
```

Here are links to the diffs for easier reviewing:

* [Syntax changes from plain JS to TS](https://github.com/cossacklabs/themis/pull/792/commits/9336d391f582e1b62d9ceec84670b4cb6bed1343)

You can simple review the commit referenced above and then only review the changes to the package.json and build files.

## Checklist

- [x] Change is covered by automated tests
- [x] Benchmark results are attached (if applicable)
- [x] The [coding guidelines] are followed
- [ ] Public API has proper documentation
- [ ] Example projects and code samples are up-to-date (in case of API changes)
- [ ] Changelog is updated (in case of notable or breaking changes)

### CODE REVIEW

**Comment by maxammann** [comment]
I'm not sure where to change example or where examples exist. The api only changed ìnitialised`to ìnitialize()`

File: src/wrappers/themis/wasm/src/libthemis.d.ts
Comment (maxammann) [inline src/wrappers/themis/wasm/src/libthemis.d.ts:1]: This file was crafted manually and describes the type of libthemis.js

File: src/wrappers/themis/wasm/src/secure_cell_context_imprint.ts
Comment (maxammann) [inline src/wrappers/themis/wasm/src/secure_cell_context_imprint.ts:62]: In a future PR I would like to remove all usages of `coerceToBytes` and demand a Uint8Array to be passed to the library. That would simplify the code in many places.

File: src/wrappers/themis/wasm/test/test.js
Comment (maxammann) [inline src/wrappers/themis/wasm/test/test.js:12]: I left test.js as is. That way I could verify that I did not break anything.

File: src/wrappers/themis/wasm/test/test.js
Comment (maxammann) [inline src/wrappers/themis/wasm/test/test.js:39]: `undefined` means no argument in TS and ES6. Therefore it is an valid argument.

File: src/wrappers/themis/wasm/wasmthemis.mk
Comment (maxammann) [inline src/wrappers/themis/wasm/wasmthemis.mk:29]: That is needed for web. This allows me to configure the path to the wasm file.

**Comment by maxammann** [comment]
I think the RustThemis failure is not introduced in this PR. For some reason Node 10 fails because of a timeout when initializing wasm. Not sure why this happens...

File: src/wrappers/themis/wasm/src/secure_keygen.ts
Comment (maxammann) [inline src/wrappers/themis/wasm/src/secure_keygen.ts]: I had to remove the extends. Officially it is not supported in ES5 to extend buildins. I want to target ES5 instead of ES6 by default to be compatible with the majority of browsers.

Comment (maxammann) [reply src/wrappers/themis/wasm/src/secure_keygen.ts]: This led to some minor changes in tests.

File: src/wrappers/themis/wasm/test/test.js
Comment (maxammann) [inline src/wrappers/themis/wasm/test/test.js:23]: instanceof is also not supported in ES5. Currently tests run with ES5 to catch comparability issues.

**Comment by maxammann** [comment]
Next steps:
* Remove tests which check for types errors
* Remove all exceptions which are thrown because of wrong types. Typescript handles this now.
* Remove legacy `arguments.length == ` handling
* Remove `coerceBytes`

**Comment by vixentael** [comment]
Wow, that's a lot 🔥

**Comment by maxammann** [comment]
> Wow, that's a lot 🔥

Yeah I tried to do minimal changes. But a rewrite is a rewrite :D

The other changes can follow in other prs. They are optional and just code quality improvements.

**Review by ilammy — COMMENTED** [review]
Overall, I'm hugely positive for this changeset. More type-level static guarantees with acceptable usability is better. I'm especially happy that someone actually experienced with JavaScript gave at look at it. I know how tedious it might have been, so props from me, Max! ❤️

**Comment by ilammy** [comment]
Also, a comment from the operational point of view for @shadinua: WasmThemis publishing should not really change much with all of this. You can still just run `make wasmthemis` to build a `wasm-themis.tgz` to be uploaded to npm and that's it. The build process gets slower though, as now it needs to download more stuff from the Internet, transpile TypeScript into JavaScript, etc.

File: src/wrappers/themis/wasm/src/index.ts
Comment (maxammann) [reply src/wrappers/themis/wasm/src/index.ts]: I agree, I already throught about throwing in prettier.

Pretter works across IDEs and is the defacto standard for formating. I add this to the todo list.

File: src/wrappers/themis/wasm/package.json
Comment (maxammann) [reply src/wrappers/themis/wasm/package.json]: I think we should change the version of the types if we upgrade emscripten. I am not sure who creates the types but i hope it is somebody who is involved in the emscripten project.

The cool thing is, if we upgrade types here and there were api changes, then we will get compile time errors.

File: src/wrappers/themis/wasm/test/test.js
Comment (maxammann) [reply src/wrappers/themis/wasm/test/test.js:73]: Yes I will put this test back. Kind of did not know what to do and removed it accidentally.

Comment (maxammann) [reply src/wrappers/themis/wasm/test/test.js:73]: Basically what we are doing in wasm themis is runtime type checking: https://learning-notes.mistermicheels.com/javascript/typescript/runtime-type-checking/

There are a few proposals in that post. One which we are already doing with `instanceof` checks. There is also a typecript runtime library which is sadly still experimental.

I'm not really sure what to do. I think it is a trade-off between cleaner code + less boiler plate and more robust + foolproove usage when using plain JS without typechecking.

I'd like to leave this decision to you. From my point of view it is better to have cleaner and simpler code and use typescript. Creating an application without typescript and still claiming that some security properties hold is at least alarming.

File: src/wrappers/themis/wasm/test/test.js
Comment (maxammann) [reply src/wrappers/themis/wasm/test/test.js:46]: >   @maxammann, is my reasoning correct so far?

Perfectly correct!

> 
>   However, I believe it's not be nice to snatch the rag from existing users and remove `initialized` entirely. Would it be possible for `initialized` and `initialize()` to coexist for a while?
> 

I will wee whether that is possible. I can not start loading themis as soon as this library is importet as you may want to configure the path to the wasm binary (required for web and webpack).

> * Changing the base type of private/public keys.

Supporting es5 is not possible with that extends syntax unfortunately. I had to support and target es5 such that the integration CI passes. So I think we should support it. The reason why it worked before is that without a compilation it somehow worked as node supports a mix between es5 and es6. Some browsers do not and only support es5. Therefore this kind of fixed bugs for users of older browsers.


BUT: We could drop support for ES5 actually because every browser which supports webassambly should already support ES6. This is also the reason why we do not need to support IE11. Webassambly does not work there in the first place.

Comment (ilammy) [reply src/wrappers/themis/wasm/test/test.js:46]: > Perfectly correct!

Whew 😌 Thanks for confirming.

> I will wee whether that is possible. I can not start loading themis as soon as this library is importet as you may want to configure the path to the wasm binary (required for web and webpack).

We have not really considered – and tested – the web usage. Knows users of WasmThemis are all on Node.js.

I remember doing some manual tests and that it kinda worked. But I'm no web developer so I believe no one so far has tested WasmThemis with proper web apps. Hence, I wouldn't be really surprised that it does not work out of the box. If something is necessary for proper web support – go ahead, you have my blessing to add whatever is needed 👍

It would be nice to keep the `initialized` promise around – I'm sorry, I don't have any ideas right now – but it it's not possible, then, well, WasmThemis users are getting a flag day. This one should not be too painful to work around.

> BUT: We could drop support for ES5 actually because every browser which supports webassambly should already support ES6.

Uh-huh... I faintly remember this reasoning behind setting the minimal supported version to ES6 and using all those fancy things.

Personally, I'm fine with ES5 support being added, but probably not at an expense of backwards-incompatible changes to the API which are caused by ES5 lack of features. If there was a way to give ES5 users one code and ES6 gets a different one – that would be great, but browsers don't work like that: it's easier to serve the same version to everyone without trying to guess what their browser might or might not support. So the approach would be something like "use some magic transpiler to convert ES6 code into ES5-compatible code".

**Comment by ilammy** [comment]
FYI, @maxammann, the fix for Rust build has been merged into master, so you can pull the updated version and merge it into your branch to turn the build green. Just a merge commit with `git checkout typescript && git merge master`, not rebases or anything like that (the PR branch is going to get squashed anyway).

**Comment by maxammann** [comment]
@ilammy any update on this one? :)

**Comment by ilammy** [comment]
@maxamman, I'm sorry, it seems that high-context culture has some effects on me and I kinda mishandled this.

It looks like we ended up playing this silly game where everybody expects other people to do something, but they never tell that explicitly and the other party does not read the air. And so weeks pass without anyone doing anything while everybody is under impression something is being done and it just takes time. From my point of view, open-source contributors are volunteers and all, they have other stuff to keep them busy. From your point of view, I think, maintainers might look the same.

I'm biased for merging this to make future work easier. However, this will block the 0.17.0 release until the backwards compatibility issues outlined in https://github.com/cossacklabs/themis/pull/792#discussion_r602680678 are resolved (one way or another: either we avoid them, or we assert them). I'm not aware of time constraints for this release, so I guess I'd go for making it when it's done rather than postponing TypeScript rework for some later release.

@vixentael, what's your take on this?

Again, @maxammann, I'm sorry for not setting my expectations straight :pray: I was under impression that you were looking into those issues, it just takes time.

**Review by ilammy — APPROVED** [review]

**Comment by maxammann** [comment]
> I'm sorry, it seems that high-context culture has some effects on me and I kinda mishandled this.

Absolutely no problem, no mishandling happened in my opinion :)

Actually, I did not see your last comment in our discussion. As far as I can see I have two open things:

- Continue to support `initialized` promise
- Drop ES5 support because all wasm supporting browsers also support ES6 and therefore also revert the changes about extending `ArrayBuffer`

**Comment by vixentael** [comment]
Thank you for your patience @maxammann! 


## Scenario 1

> Continue to support initialized promise

Would be nice if we can!

> Drop ES5 support 

I'm ok with this., it's old :)


If we can support old `initialized` promise OR write deprecation notice on that one, then I think we can merge TS support to master and release with next Themis release.

## Scenario 2

If we can't support old `initialized` promise, then I suggest merging these changes to a separate branch `master/typescript`, and aim at releasing `0.14.0` with breaking changes (add initialize(); drop ES5) :)

I'd love to support backwards compatibility, but I understand that WASM Themis is typically used in desktop and web apps, which move fast. Soo... sorry devs? We will describe breaking changes as detailed as possible.

---

@maxammann do you think it's possible to support `initialized` promise or to write some deprecation/breaking change notice in code to let devs know that they should use `initialize` now?

**Comment by ilammy** [comment]
Alrighty. @maxammann, I've retargeted your PR onto a new branch – `wasm-typescript` – and merged it there. Please target your future TypeScript PRs for that branch. Once TypeScript support is stabilized and ready for release, the `wasm-typescript` branch will be merged into `master`.

This way if Themis really needs a release, we can provide it without potentially introducing a breaking change. Hopefully, we won't need this and TypeScript will be in there in Themis 0.14.

**Comment by maxammann** [comment]
Thanks! Right now I don't have time to continue but soon I will :)

Just ping me when there are urgent changes for wasm-themis.

**Pull request merged**
